### PR TITLE
feat: eth: Avoid StateCompute in EthTxnReceipt lookup

### LINKED ISSUE #10418: EthAPI: Optimize EthGetTransactionReceipt
It currently performs 2 unbounded lookbacks from current head -- one to "find" the message, and a second to "replay" it. The second lookback can be entirely avoided by providing the tipset from the "find" step to StateReplay.

This is one of (likely) many perf improvements that can be made to the EthAPI, but was flagged as being an especially important method.

### PR DESCRIPTION
**Pull request by Stebalien**
## Related Issues

fixes #10418

## Proposed Changes

Instead:

1. Use the receipt from the message search.
2. Re-compute the gas fees that would have been charged.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] New features have usage guidelines and / or documentation updates in
  - [x] [Lotus Documentation](https://lotus.filecoin.io)
  - [x] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -2075,25 +2071,34 @@ func newEthTxReceipt(ctx context.Context, tx ethtypes.EthTx, replay *api.InvocRe
...
 	}
 
-	receipt.GasUsed = ethtypes.EthUint64(replay.MsgRct.GasUsed)
+	receipt.GasUsed = ethtypes.EthUint64(lookup.Receipt.GasUsed)
 
 	// TODO: handle CumulativeGasUsed
 	receipt.CumulativeGasUsed = ethtypes.EmptyEthInt
 
-	effectiveGasPrice := big.Div(replay.GasCost.TotalCost, big.NewInt(replay.MsgRct.GasUsed))
+	// TODO: avoid loading the tipset twice (once here, once in the when we convert the message to a txn)
```
Comment (Stebalien) [inline node/impl/full/eth.go:2085]: Not critical right now (it's guaranteed to be cached). It's just a bit annoying.

**Review by Stebalien — COMMENTED** [review]
@raulk's changes LGTM, but I need a cross review from him to merge because I made the PR.

File: node/impl/full/eth.go (hunk 1 above)
Comment (arajasek) [reply node/impl/full/eth.go:2085]: Agreed, but grammar of the comment is wonky (`once in the when we convert`).

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -2075,25 +2071,34 @@ func newEthTxReceipt(ctx context.Context, tx ethtypes.EthTx, replay *api.InvocRe
...
+	ts, err := cs.GetTipSetFromKey(ctx, lookup.TipSet)
+	if err != nil {
+		return api.EthTxReceipt{}, err
```
Comment (arajasek) [inline node/impl/full/eth.go:2088]: Can we chain error here? It makes debugging a lot easier.

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -2075,25 +2071,34 @@ func newEthTxReceipt(ctx context.Context, tx ethtypes.EthTx, replay *api.InvocRe
...
+	}
+
+	baseFee := ts.Blocks()[0].ParentBaseFee
+	gasOutputs := vm.ComputeGasOutputs(lookup.Receipt.GasUsed, int64(tx.Gas), baseFee, big.Int(tx.MaxFeePerGas), big.Int(tx.MaxPriorityFeePerGas), true)
+	totalBurnt := big.Sum(gasOutputs.BaseFeeBurn, gasOutputs.MinerTip, gasOutputs.OverEstimationBurn)
```
[the remark this answers is not in the document]
Comment (arajasek) [reply node/impl/full/eth.go:2093]: Yes, though I'd suggest renaming this variable (it's not the total _burnt_, just the total spent).

**Review by arajasek — APPROVED** [review]

**Pull request merged**